strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,106 @@ def context(self, *args, copies=1, shift="default"):
         self.loop_vars = self.params(*[f"p{i}" for i in range(len(args))])
         return self
 
+    def compile(self, *, device=None, compiler=None):
+        """Compile the time-domain program given a Strawberry Fields photonic hardware device specification.
+        At this stage the compilation is simply a check that the program matches the device.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): device specification object to use for
+                program compilation
+            compiler (str, ~strawberryfields.compilers.Compiler): Compiler name or compile strategy
+                to use. If a device is specified, this overrides the compile strategy specified by
+                the hardware :class:`~.DeviceSpec`. For :class:`~.TDMProgram` this should be set to
+                None.
+
+        Returns:
+            Program: compiled program
+        """
+        if compiler == "gaussian":
+            return super().compile(device=device, compiler=compiler)
+
+        # The next two if statements need to be updated, in particular with the name of the programs
+        if compiler == "tdm_one_loop" and device == "tdm_one_loop_device":
+            # Do something clever
+            compiler = "td2"  # Devious but it will work, Josh said
+            # Do not add a return statement
+
+        if compiler == "td2" and device is not None:
+            device_layout = bb.loads(device.layout)
+
+            # First check: the gates are in the correct order
+            program_gates = [cmd.op.__class__.__name__ for cmd in self.rolled_circuit]
+            device_gates = [op["op"] for op in device_layout.operations]
+            if device_gates != program_gates:
+                raise CircuitError(
+                    "The gates or the order of gates used in the Program is incompatible with the device '{}' ".format(
+                        device.target
+                    )
+                )
+
+            # Second check: the gates act on the correct modes
+            program_modes = [[r.ind for r in cmd.reg] for cmd in self.rolled_circuit]
+            device_modes = [op["modes"] for op in device_layout.operations]
+            if program_modes != device_modes:
+                raise CircuitError(
+                    "Program cannot be used with the device '{}' "
+                    "due to incompatible mode ordering.".format(device.target)
+                )
+
+            # Third check: the parameters of the gates are valid
+            gate_params_ranges = device.gate_parameters
+            # We will loop over the different operations in the device specification
+
+            for i, operation in enumerate(device_layout.operations):
+                # We obtain the name of the parameter(s)
+                param_names = operation["args"]
```

Comment:
`operation["args"]` will only be a list of names/strings if the parameters are named `p0`, `p1`, etc. What if we want to add a fixed gate to the device spec layout that has hard-coded parameters. Could this ever be the case? 

If so, it might be worth adding a check that this is a string and, if not, simply check if `self.rolled_circuit[i].op.p[k] == operation["args"][k]`. Otherwise, the line `param_range = device.gate_parameters[param_name]` below will raise a `KeyError`.

Reply:
I see. I did not know about this feature of BB. Sounds like a good addition!

Reply:
It's more that the `operation["args"]` have always strictly returned values before. It's just the TDM addition that added the ability to let them be parameters (named `p0`, `p1`, etc.), and thus letting them be names/strings instead.